File: mobile/lib/shared/events/bus.dart
import 'dart:async';

import 'package:flutter_riverpod/flutter_riverpod.dart';

/// In-app cross-feature event bus, mirroring frontend-v2
/// `shared/events/bus.ts`. Features never import each other; the sender
/// emits and whoever cares listens. The one live event today (as on web):
/// `workbench.open` `{kind: review|terminal|files, file?}`.
class AppEvent {
  const AppEvent(this.type, [this.payload = const {}]);

  final String type;
  final Map<String, Object?> payload;
}

class AppEventBus {
  final _controller = StreamController<AppEvent>.broadcast();

  Stream<AppEvent> get stream => _controller.stream;

  Stream<AppEvent> on(String type) =>
      _controller.stream.where((e) => e.type == type);

  void emit(String type, [Map<String, Object?> payload = const {}]) {
    _controller.add(AppEvent(type, payload));
  }
}

final appEventBusProvider = Provider<AppEventBus>((ref) => AppEventBus());

File: frontend-v2/src/shared/events/bus.ts
// In-app event bus for cross-feature signals (ENGINEERING_SPEC §4.2 事件解耦).
// Chat emits "open review" — workbench listens. Neither imports the other.
type AppEventMap = {
  "workbench.open": { kind: "review" | "terminal" | "browser" | "files" | "cron"; file?: string }
}

type AppEventName = keyof AppEventMap
type Handler<E extends AppEventName> = (data: AppEventMap[E]) => void

const handlers = new Map<string, Set<Handler<AppEventName>>>()

export function onAppEvent<E extends AppEventName>(event: E, handler: Handler<E>): () => void {
  const set = handlers.get(event) ?? new Set()
  set.add(handler as Handler<AppEventName>)
  handlers.set(event, set)
  return () => set.delete(handler as Handler<AppEventName>)
}

export function emitAppEvent<E extends AppEventName>(event: E, data: AppEventMap[E]): void {
  handlers.get(event)?.forEach((h) => h(data))
}
